src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    ///
+    /// May return `Err` for zero-sized `T` or `n == 0`.
+    ///
+    /// Always returns `Err` on arithmetic overflow.
+    unsafe fn alloc_array<T>(&mut self, n: usize) -> Result<Unique<T>, AllocErr>
+        where Self: Sized
+    {
+        match Layout::array::<T>(n) {
+            Some(ref layout) if layout.size() > 0 => {
+                self.alloc(layout.clone())
+                    .map(|p| {
+                        Unique::new(p as *mut T)
+                    })
+            }
+            _ => Err(AllocErr::invalid_input("invalid layout for alloc_array")),
+        }
+    }
+
+    /// Reallocates a block previously suitable for holding `n_old`
+    /// instances of `T`, returning a block suitable for holding
+    /// `n_new` instances of `T`.
+    ///
+    /// Captures a common usage pattern for allocators.
+    ///
+    /// The returned block is suitable for passing to the
+    /// `alloc`/`realloc` methods of this allocator.
+    ///
+    /// May return `Err` for zero-sized `T` or `n == 0`.
+    ///
+    /// Always returns `Err` on arithmetic overflow.
+    unsafe fn realloc_array<T>(&mut self,
+                               ptr: Unique<T>,
+                               n_old: usize,
+                               n_new: usize) -> Result<Unique<T>, AllocErr>
+        where Self: Sized
+    {
+        match (Layout::array::<T>(n_old), Layout::array::<T>(n_new), ptr.as_ptr()) {
+            (Some(ref k_old), Some(ref k_new), ptr) if k_old.size() > 0 && k_new.size() > 0 => {
+                self.realloc(ptr as *mut u8, k_old.clone(), k_new.clone())
+                    .map(|p|Unique::new(p as *mut T))
+            }
+            _ => {
+                Err(AllocErr::invalid_input("invalid layout for realloc_array"))
+            }
+        }
+    }
+
+    /// Deallocates a block suitable for holding `n` instances of `T`.
+    ///
+    /// Captures a common usage pattern for allocators.
+    unsafe fn dealloc_array<T>(&mut self, ptr: Unique<T>, n: usize) -> Result<(), AllocErr>
+        where Self: Sized
+    {
+        let raw_ptr = ptr.as_ptr() as *mut u8;
+        match Layout::array::<T>(n) {
+            Some(ref k) if k.size() > 0 => {
+                Ok(self.dealloc(raw_ptr, k.clone()))
+            }
+            _ => {
+                Err(AllocErr::invalid_input("invalid layout for dealloc_array"))
+            }
+        }
+    }
+
+    // UNCHECKED METHOD VARIANTS
+
+    /// Returns a pointer suitable for holding data described by
+    /// `layout`, meeting its size and alignment guarantees.
+    ///
+    /// The returned block of storage may or may not have its contents
+    /// initialized. (Extension subtraits might restrict this
+    /// behavior, e.g. to ensure initialization.)
+    ///
+    /// Returns `None` if request unsatisfied.
+    ///
+    /// Behavior undefined if input does not meet size or alignment
+    /// constraints of this allocator.
```

Comment:
So to be clear, anything that would not result in an `Unsupported` error in `alloc` is valid to pass here? If that's the case, that would imply that, if the allocator doesn't support zero-sized allocations, `alloc` is required to check for them and return `Unsupported` (which makes sense, but also seems inefficient since `alloc` is the common case - a random generic library can never call `alloc_unchecked` because it never knows for sure that the layouts it wants are supported).